x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,595 @@
+                    }
+                }
+                successfulShards++;
+            }
+        }
+        int size = Math.min(request.size(), combinedResults.size());
+        List<String> terms = new ArrayList<>(size);
+        TermCount[] sortedCombinedResults = combinedResults.values().toArray(new TermCount[0]);
+        if (request.sortByPopularity()) {
+            // Sort by doc count descending
+            Arrays.sort(sortedCombinedResults, new Comparator<TermCount>() {
+                public int compare(TermCount t1, TermCount t2) {
+                    return Long.compare(t2.getDocCount(), t1.getDocCount());
+                }
+            });
+        } else {
+            // Sort alphabetically
+            Arrays.sort(sortedCombinedResults, new Comparator<TermCount>() {
+                public int compare(TermCount t1, TermCount t2) {
+                    return t1.getTerm().compareTo(t2.getTerm());
+                }
+            });
+        }
+
+        for (TermCount term : sortedCombinedResults) {
+            terms.add(term.getTerm());
+            if (terms.size() == size) {
+                break;
+            }
+        }
+//        long timeTook = System.currentTimeMillis() - request.taskStartTimeMillis;
+//        System.err.println("Took "+timeTook+" ms");
+        return new TermEnumResponse(terms, shardsResponses.length(), successfulShards, failedShards, shardFailures, timedOut);
+    }
+
+    static class TermCountPriorityQueue extends PriorityQueue<TermCount> {
+
+        TermCountPriorityQueue(int maxSize) {
+            super(maxSize);
+        }
+
+        @Override
+        protected boolean lessThan(TermCount a, TermCount b) {
+            return a.getDocCount() < b.getDocCount();
+        }
+
+    }
+
+    protected ShardTermEnumResponse shardOperation(ShardTermEnumRequest request, Task task) throws IOException {
+        List<TermCount> termsList = new ArrayList<>();
+        String error = null;
+        
+        long timeout_millis = request.timeout();
+        long scheduledEnd = request.shardStartedTimeMillis() + timeout_millis;
+
+        // Check we haven't just arrived on a shard and time is up already.
+        if (System.currentTimeMillis() > scheduledEnd) {
+            return new ShardTermEnumResponse(request.shardId(), termsList, error, true);
+        }        
+        
+        // Like TransportFieldCapabilitiesAction - fail fast if the filter excludes this index.
+        if (canMatchShard(request) == false) {
+            return new ShardTermEnumResponse(request.shardId(), termsList, error, false);
+        }
+        ShardSearchRequest shardSearchLocalRequest = new ShardSearchRequest(
+            request.shardId(),
+            request.taskStartedTimeMillis(),
+            request.filteringAliases()
+        );
+        SearchContext searchContext = searchService.createSearchContext(shardSearchLocalRequest, SearchService.NO_TIMEOUT);
```

Comment:
You don't need a `SearchContex` here, you should use a`QueryShardContext` instead like we do in the can match phase. Also note that you could use the same `QueryShardContext` for the canMatchShard above, `SearchService#queryStillMatchesAfterRewrite` was added for this purpose.